book/src/json-output.md

```diff
@@ -0,0 +1,157 @@
+{"data":"warn","host_timestamp":1643113389707290115,"level":"WARN","location":{"file":"src/bin/levels.rs","line":12,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"1"}
+{"data":"debug","host_timestamp":1643113389707299759,"level":"DEBUG","location":{"file":"src/bin/levels.rs","line":13,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"2"}
+{"data":"error","host_timestamp":1643113389707306961,"level":"ERROR","location":{"file":"src/bin/levels.rs","line":14,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"3"}
+{"data":"println","host_timestamp":1643113389707313290,"level":null,"location":{"file":"src/bin/levels.rs","line":15,"module_path":{"crate_name":"levels","modules":[],"function":"__cortex_m_rt_main"}},"target_timestamp":"4"}
+```
+
+🤔: That seems convenient, but what is this schema version in the first line?
+
+It indicates the version of the json format you are using. `probe-run` will always output it as a header at the beginning of each stream of logs. We anticipate that the format will slightly change while `probe-run` and `defmt` evolve. Using this version you always know which revision is in use and can act upon that.
+
+🤗: Thank you for the explanation!
+
+## JsonFrame
+
+The [`defmt-json-schema`] crate provides a data transfer object `struct JsonFrame` which is used to serialize the log frame to json and can be used to deserialize it back to rust as well.
+
+The `JsonFrame` is defined like this ...
+
+```rust
+pub struct JsonFrame {
+    pub data: String,
+    pub host_timestamp: i64,
+    pub level: Option<log::Level>,
+    pub location: Location,
+    pub target_timestamp: String,
+}
+
+pub struct Location {
+    pub file: Option<String>,
+    pub line: Option<u32>,
+    pub module_path: Option<ModulePath>,
+}
+
+pub struct ModulePath {
+    pub crate_name: String,
+    pub modules: Vec<String>,
+    pub function: String,
+}
+
+# mod log { pub struct Level; }
+```
+
+... which results in following json schema ...
+
+```json
+{
+    "data": "string",
+    "host_timestamp": "number",
+    "level": null | "string",
+    "location": {
+        "file": null | "string",
+        "line": null | "number",
+        "module_path": null | {
+            "crate_name": "string",
+            "modules": ["string"],
+            "function": "string"
+        }
+    },
+    "target_timestamp": "string"
+}
+```
+
+... which looks like this, for example:
+
+```json
+{
+    "data": "Hello, world!",
+    "host_timestamp": 1643110723881550007,
+    "level": "INFO",
+    "location": {
+        "file": "src/bin/hello.rs",
+        "line": 8,
+        "module_path": {
+            "crate_name": "hello",
+            "modules": [],
+            "function": "__cortex_m_rt_main"
+        }
+    },
+    "target_timestamp": "1"
+}
+```
+
+## SchemaVersion
+
```

Comment:
What's this for, or how would you use it?

Reply:
Are you talking about the example of the JSON output, or the SchemaVersion?

Reply:
The SchemaVersion. It wasn't clear what "can be used to deserialise the JSON back to Rust" had to do with a struct containing a field called `schema_version`. So maybe an example would help?

Reply:
Oh wait, I think I get it. Maybe add something like:

> If you wish to deserialise the entire file back into a Rust program, you will need to be able to decode the `SchemaVersion` object at the start of the file, as well as the `JsonFrame` objects. To do that, we supply a `SchemaVersion` struct in .... which you can use by ....

Reply:
I've reformulated it. What do you think now?

Reply:
Nice :) 👍